### PR TITLE
Adjust the Greengrass Discovery client to use its own executor service

### LINKED ISSUE #320: Greengrass DiscoveryClient does not work if a SecurityManager is installed
### Describe the bug

If a SecurityManager is installed, Greengrass Core Device discovery performed using DiscoveryClient will fail even with a permissive policy.

### Expected Behavior

Greengrass Discovery completes successfully even if a security manager is installed.

### Current Behavior

Greengrass Core Device discovery performed using DiscoveryClient fails with exception [1]

[1]
```
Exception thrown: java.util.concurrent.ExecutionException: com.google.gson.JsonIOException: Failed making field 'software.amazon.awssdk.iot.discovery.model.DiscoverResponse#GGGroups' accessible; either change its visibility or write a custom TypeAdapter for its declaring type
java.util.concurrent.ExecutionException: com.google.gson.JsonIOException: Failed making field 'software.amazon.awssdk.iot.discovery.model.DiscoverResponse#GGGroups' accessible; either change its visibility or write a custom TypeAdapter for its declaring type
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at greengrass.BasicDiscovery.getClientFromDiscovery(BasicDiscovery.java:157)
	at greengrass.BasicDiscovery.main(BasicDiscovery.java:114)
Caused by: com.google.gson.JsonIOException: Failed making field 'software.amazon.awssdk.iot.discovery.model.DiscoverResponse#GGGroups' accessible; either change its visibility or write a custom TypeAdapter for its declaring type
	at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:22)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:158)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:101)
	at com.google.gson.Gson.getAdapter(Gson.java:501)
	at com.google.gson.Gson.fromJson(Gson.java:990)
	at com.google.gson.Gson.fromJson(Gson.java:929)
	at software.amazon.awssdk.iot.discovery.DiscoveryClient.lambda$discover$0(DiscoveryClient.java:97)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
	at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1596)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
Caused by: java.security.AccessControlException: access denied ("java.lang.reflect.ReflectPermission" "suppressAccessChecks")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
	at java.security.AccessController.checkPermission(AccessController.java:886)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
	at java.lang.reflect.AccessibleObject.setAccessible(AccessibleObject.java:128)
	at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:19)
	... 12 more
```

### Reproduction Steps

Build a standalone JAR file that executes https://github.com/aws/aws-iot-device-sdk-java-v2/blob/main/samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java as main class. This can obtained for example by adding the snippet [1] to the build -> plugins section of https://github.com/aws/aws-iot-device-sdk-java-v2/blob/main/samples/Greengrass/pom.xml.

Run the produced jar file with the following command:

`java -Djava.security.manager -Djava.security.policy=security.policy -jar ./Greengrass-1.0-SNAPSHOT-jar-with-dependencies.jar ${DISCOVERY_SAMPLE_ARGS}`

where `DISCOVERY_SAMPLE_ARGS` is as described in https://docs.aws.amazon.com/greengrass/v2/developerguide/test-client-device-communications.html#test-client-device-communications-java and `security.policy` contains the following maximally permissive policy:

```
grant { 
      permission java.security.AllPermission;
};
```

The command should fail logging the reported exception if the`-Djava.security.manager -Djava.security.policy=security.policy` arguments are provided and succeed if they are removed.

[1]
```
<plugin>
    <artifactId>maven-assembly-plugin</artifactId>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>single</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <archive>
            <manifest>
                <mainClass>greengrass.BasicDiscovery</mainClass>
            </manifest>
        </archive>
        <descriptorRefs>
            <descriptorRef>jar-with-dependencies</descriptorRef>
        </descriptorRefs>
    </configuration>
</plugin>
```

### Possible Solution

The issue is likely caused by the fact that

https://github.com/aws/aws-iot-device-sdk-java-v2/blob/472d0ecd511b3e31ee65882636dc91cbcd5d1264/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java#L97

is executed through [CompletableFuture.supplyAsync()](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#supplyAsync-java.util.function.Supplier-) using the Fork Join Pool common pool.

Quoting from [ForkJoinPool Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html), _If a [SecurityManager](https://docs.oracle.com/javase/8/docs/api/java/lang/SecurityManager.html) is present and no factory is specified, then the default pool uses a factory supplying threads that have no [Permissions](https://docs.oracle.com/javase/8/docs/api/java/security/Permissions.html) enabled._.

### Additional Information/Context

_No response_

### SDK version used

main branch cloned from https://github.com/aws/aws-iot-device-sdk-java-v2/commit/1a2e1b5d82eab57e8365ca91883dc1ad9f3a32c3

### Environment details (OS name and version, etc.)

Raspberry PI OS, openjdk version "1.8.0_312" OpenJDK Runtime Environment (build 1.8.0_312-8u312-b07-1+rpi1-b07) OpenJDK Client VM (build 25.312-b07, mixed mode)

### PR DESCRIPTION
**Pull request by TwistedTwigleg**
*Issue #, if available:*

Fixes #320

*Description of changes:*

As part of #426, I was investigating why I was getting warnings saying threads were not being closed and it was resulting in a roughly minute long delay at the end of the sample, though the sample itself works fine and all the code executes. 

The warning I got with the sample running locally in #426 was:

~~~
[WARNING] thread Thread[#33,ForkJoinPool.commonPool-worker-1,5,greengrass.BasicDiscovery] was interrupted but is still alive after waiting at least 15000msecs
[WARNING] thread Thread[#33,ForkJoinPool.commonPool-worker-1,5,greengrass.BasicDiscovery] will linger despite being asked to die via interruption
[WARNING] thread Thread[#34,ForkJoinPool.commonPool-worker-2,5,greengrass.BasicDiscovery] will linger despite being asked to die via interruption
[WARNING] NOTE: 2 thread(s) did not finish despite being asked to  via interruption. This is not a problem with exec:java, it is a problem with the running code. Although not serious, it should be remedied.
~~~

Which was a bit concerning as it seemed some thread was lingering after the sample was finished.

After a bunch of researching, poking around the code, and looking around, I concluded the issue seemed to be due to the common thread pool being used and not being properly cleaned up by default because it uses daemon/non-daemon threads and using the wrong type can lead the JVM to finish before the thread, orphaning the thread (if my admitted limited understanding is correct).
By default, `supplyAsync()` uses the default common thread pool, which has a tendency to have this issue to occur, apparently from what I gathered. While I didn't find an exact case that was showing the warnings I was seeing or was like our code with `supplyAsync`, the overall common advice I found while researching is to make and manage a thread pool, calling `shutdown` when finished, which frees the threads in question.

I made the adjustment in this PR, making the code manage its own single thread pool just for this single purpose. It also adds a new configuration option to set the executor manually, if desired, in the discovery client config. If set manually, the customer is expected to shut the executor down when finished. (this is also documented)

I tested and with this change, the minute long delay and warnings were gone! Everything still functions properly too.

Looking at the issue in #320, it seemed similar and so I tested it, and it seems to work there now too! I confirmed that without the change the issue occurs and you get an exception, and with the change, the issue is not present. Based on really quick research into that issue, it seems that the default common thread pool does NOT carry over the permissions of the executing Jar, but if you make one manually, it does properly inherit the security permissions.

**TLDR**: The delay that I was seeing in #426 is fixed by having the discovery client make and destroy its own thread pool in an executor service, and it happened to fix the SecurityManager issue as well.

___________

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClientConfig.java [hunk 1]
```diff
@@ -164,6 +169,27 @@ public String getGGServerName() {
         return ggServerName;
     }
 
+    /**
+     * @return the executor set for this discover client, if one is set.
```
Comment (sbSteveK) [inline sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClientConfig.java:173]: Semantics but it returns whether an executor is set or not. This should probably read that it returns the set executor or null.

File: sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java [hunk 2]
```diff
@@ -120,5 +137,9 @@ public void close() {
         if(httpClientConnectionManager != null) {
             httpClientConnectionManager.close();
         }
+        if (cleanExecutor == true && executorService != null) {
```
Comment (sbSteveK) [inline sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java:140]: I don't think we need the executorService != null check unless there's some way that the executorService can be shutdown and made null outside of this close() after having been created specifically for this client.

**Review by xiazhvera — APPROVED** [review]
